banzai/images.py

```diff
@@ -1,272 +1,766 @@
+    @abc.abstractmethod
+    def to_fits(self) -> Union[fits.HDUList, list]:
+        pass
 
-        if data_tables is None:
-            data_tables = {}
 
-        if extension_headers is None:
-            extension_headers = []
+class HeaderOnly(Data):
+    def __init__(self, meta: Union[dict, fits.Header]):
+        super().__init__(data=np.zeros(0), meta=meta)
 
-        if filename is not None:
-            data, header, bpm, extension_headers = fits_utils.open_image(filename)
-            if '.fz' == filename[-3:]:
-                filename = filename[:-3]
-            self.filename = os.path.basename(filename)
+    def to_fits(self):
+        return fits.HDUList([fits.ImageHDU(data=None, header=self.meta)])
 
-        self.data = data
-        self.data_tables = data_tables
-        self.header = header
-        self.bpm = bpm
 
-        self.extension_headers = extension_headers
+class CCDData(Data):
+    def __init__(self, data: Union[np.array, Table], meta: Union[dict, fits.Header],
+                 mask: np.array = None, name: str = '', uncertainty: np.array = None):
+        super().__init__(data=data, meta=meta, mask=mask, name=name)
+        if uncertainty is None:
+            uncertainty = self.read_noise * np.ones(data.shape, dtype=data.dtype) / self.gain
+        self.uncertainty = self._init_array(uncertainty)
 
-        self.request_number = header.get('REQNUM')
-        self.instrument, self.site, self.camera = self._init_instrument_info(runtime_context)
+    def __getitem__(self, section):
+        """
+        Return a new CCDData object with the given section of data
+        :param section: needs to be  in data coords
+        :return:
+        """
+        return self.trim(trim_section=section)
+
+    def __imul__(self, value):
+        self.data *= value
+        self.meta['SATURATE'] *= value
+        self.meta['GAIN'] *= value
+        self.meta['MAXLIN'] *= value
+        return self
+
+    def __idiv__(self, other):
+        self.__imul__(1.0 / other)
+        return self
+
+    def to_fits(self):
+        data_hdu = fits.ImageHDU(data=self.data, header=fits.Header(self.meta))
+        bpm_extname = self.extension_name + 'BPM'
+        mask_hdu = fits.ImageHDU(data=self.mask, header=fits.Header({'EXTNAME': bpm_extname}))
+        uncertainty_extname = self.extension_name + 'ERR'
+        uncertainty_hdu = fits.ImageHDU(data=self.uncertainty, header=fits.Header({'EXTNAME': uncertainty_extname}))
+        hdulist = fits.HDUList([data_hdu, mask_hdu, uncertainty_hdu])
+        return hdulist
+
+    def __del__(self):
+        super().__del__()
+        del self.uncertainty
+
+    def __isub__(self, value):
+        if isinstance(value, CCDData):
+            self.data -= value.data
+            self.uncertainty = np.sqrt(value.uncertainty * value.uncertainty + self.uncertainty * self.uncertainty)
+            self.mask |= value.mask
+        else:
+            self.data -= value
+        return self
```

Comment:
Do both __isub__ and __sub__ need to be defined for image? If a and b were CCDData In a little example I did, it seems like __sub__ is sufficient for getting both x = a-b  to work and a-=b to work. Or is __isub__ for something other than defining a-=b ?